### PR TITLE
Add Noise Protocol Support

### PR DESCRIPTION
**Pull request by igor-egorov**
Works!

Tested configurations:
go server & c++ client
c++ server & go client
c++ server & c++ client
(go server & go client)

Many thanks to @lebdron for co-debugging!

https://github.com/libp2p/specs/tree/master/noise
https://noiseexplorer.com/patterns/XX/
http://noiseprotocol.org/noise.html

### CODE REVIEW

**Review by kamilsa — CHANGES_REQUESTED** [review]

File: include/libp2p/security/noise/crypto/noise_dh.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_DH_HPP
+#define LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_DH_HPP
+
+#include <libp2p/crypto/x25519_provider/x25519_provider_impl.hpp>
+#include <libp2p/security/noise/crypto/interfaces.hpp>
+
+namespace libp2p::security::noise {
+
+  /// X25519 DH
+  class DiffieHellmanImpl : public DiffieHellman {
```
Comment (kamilsa) [inline include/libp2p/security/noise/crypto/noise_dh.hpp:15]: It would be nicer if name of file is consistent with class name

File: include/libp2p/security/noise/crypto/noise_sha256.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+#ifndef LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_SHA256_HPP
+#define LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_SHA256_HPP
+
+#include <libp2p/crypto/sha/sha256.hpp>
+#include <libp2p/security/noise/crypto/interfaces.hpp>
+
+namespace libp2p::security::noise {
+
+  class SHA256 : public NamedHash {
```
Comment (kamilsa) [inline include/libp2p/security/noise/crypto/noise_sha256.hpp:14]: Same. Can we make consistent names of the file and class?

File: include/libp2p/security/noise/crypto/noise_ccp1305.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+#ifndef LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_CCP1305_HPP
+#define LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_NOISE_CCP1305_HPP
+
+#include <libp2p/crypto/chachapoly/chachapoly_impl.hpp>
+#include <libp2p/security/noise/crypto/interfaces.hpp>
+
+namespace libp2p::security::noise {
+  class CCP1305Impl : public AEADCipher {
```
Comment (kamilsa) [inline include/libp2p/security/noise/crypto/noise_ccp1305.hpp:13]: Name of file and class are not consistent

**Review by kamilsa — CHANGES_REQUESTED** [review]

File: include/libp2p/crypto/chachapoly.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  class ChaCha20Poly1305 {
+   public:
+    virtual ~ChaCha20Poly1305() = default;
+
+    /*
+     * NB:
+     * Descendants are to implement a constructor or other method to set the
+     * mode which crypt method should perform (encrypt or decrypt) as well as
+     * the key setup.
```
Comment (Harrm) [inline include/libp2p/crypto/chachapoly.hpp:37]: Why is this required by the interface?

File: include/libp2p/crypto/chachapoly/chachapoly_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <openssl/evp.h>
+#include <libp2p/common/logger.hpp>
+#include <libp2p/crypto/chachapoly.hpp>
+
+namespace libp2p::crypto::chachapoly {
+
+  class ChaCha20Poly1305Impl : public ChaCha20Poly1305 {
+   public:
+    ChaCha20Poly1305Impl( Key key/*, Mode default_mode*/);
```
Comment (Harrm) [inline include/libp2p/crypto/chachapoly/chachapoly_impl.hpp:17]: Commented code.

Comment (Harrm) [reply include/libp2p/crypto/chachapoly/chachapoly_impl.hpp:17]: And this could be explicit, I guess.

File: include/libp2p/crypto/hash.hpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <gsl/span>
+#include <libp2p/crypto/common.hpp>
+#include <libp2p/outcome/outcome.hpp>
+
+namespace libp2p::crypto {
+
+  using libp2p::crypto::common::HashType;
+
+  class Hash {
```
[the remark this answers is not in the document]
Comment (Harrm) [reply include/libp2p/crypto/hash.hpp:17]: I would even propose "Hasher" because it clearly processes some internal logic, while Hash sounds more like a data structure.

File: include/libp2p/crypto/x25519_provider.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * Diffie Hellman key agreement calculator
+   */
+  class X25519Provider {
+   public:
+    virtual ~X25519Provider() = default;
+
+    /**
+     * Generates a keypair using x25519 curve
```
Comment (Harrm) [inline include/libp2p/crypto/x25519_provider.hpp:32]: As a matter of terminology: AFAIK x25519 is the DH function, where the name of the underlying curve is Curve25519.

File: include/libp2p/crypto/x25519_provider.hpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    virtual outcome::result<PublicKey> derive(
+        const PrivateKey &private_key) const = 0;
+
+    /**
+     * Does dh calculation - derives shared secret
+     * @param private_key
+     * @param public_key
+     * @return dh result as bytes
+     */
+    virtual outcome::result<std::vector<uint8_t>> dh(
```
Comment (Harrm) [inline include/libp2p/crypto/x25519_provider.hpp:51]: Although Diffie-Helman is quite a famous term, I would still attempt to convince you to rename it to something more comprehensive.

File: include/libp2p/security/noise/crypto/interfaces.hpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  class NamedAEADCipher {
+   public:
+    virtual ~NamedAEADCipher() = default;
+
+    virtual std::shared_ptr<AEADCipher> cipher(Key32 key) = 0;
+
+    virtual std::string cipherName() const = 0;
+  };
+
+  class CipherSuite : public DiffieHellman,
```
Comment (Harrm) [inline include/libp2p/security/noise/crypto/interfaces.hpp:100]: Please comment.

File: include/libp2p/security/noise/crypto/message_patterns.hpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+#ifndef LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_MESSAGE_PATTERNS_HPP
+#define LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_CRYPTO_MESSAGE_PATTERNS_HPP
+
+#include <string>
+#include <vector>
+
+namespace libp2p::security::noise {
+
+  enum class MessagePattern : int {  // int required here
```
Comment (Harrm) [inline include/libp2p/security/noise/crypto/message_patterns.hpp:14]: Would be great to see short comments explaining the patterns or a link to the docs.

File: include/libp2p/security/noise/crypto/state.hpp [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    MESSAGE_TOO_LONG,
+    MESSAGE_TOO_SHORT,
+    NO_PUBLIC_KEY,
+    REMOTE_KEY_ALREADY_SET,
+  };
+
+  outcome::result<Key32> bytesToKey32(gsl::span<const uint8_t> key);
+
+  /// Provides symmetric encryption and decryption after a successful handshake
+  class CipherState {
```
Comment (Harrm) [inline include/libp2p/security/noise/crypto/state.hpp:34]: I'm not sure if I completely understand why does a Cipher_State_ encrypts and decrypts something and I believe that it would give me hard time understanding what it is without the walkthrough through the PR.

File: include/libp2p/security/noise/handshake.hpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+#include <libp2p/security/noise/insecure_rw.hpp>
+#include <libp2p/security/security_adaptor.hpp>
+
+namespace libp2p::security::noise {
+
+  std::shared_ptr<CipherSuite> defaultCipherSuite();
+
+  class Handshake : public std::enable_shared_from_this<Handshake> {
+   public:
+    explicit Handshake(
```
Comment (Harrm) [inline include/libp2p/security/noise/handshake.hpp:28]: Would it not be better to introduce some kind of a builder? This constructor is monstrous.

File: src/crypto/chachapoly/chachapoly_impl.cpp [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+      IF1(EVP_DecryptUpdate(ctx, nullptr, &len, aad.data(), aad.size()),
+          "Failed to apply additional authentication data during decryption.",
+          OpenSslError::FAILED_DECRYPT_UPDATE)
+    }
+
+    IF1(EVP_DecryptUpdate(ctx, result.data(), &len, ciphertext.data(),
+                          ciphertext.size() - 16),
+        "Ciphertext decryption failed.", OpenSslError::FAILED_DECRYPT_UPDATE)
+
+    if (1
```
Comment (Harrm) [inline src/crypto/chachapoly/chachapoly_impl.cpp:121]: Isn't it the same pattern as in the scary macro IF1?

File: src/crypto/hmac_provider/hmac_provider_impl.cpp [hunk 14]
```diff
@@ -40,38 +43,156 @@ namespace libp2p::crypto::hmac {
...
+    if (initialized_) {
+      return EVP_MD_block_size(hash_st_);
+    }
+    return 0;
+  }
+
+  void HmacProviderCtrImpl::sinkCtx(size_t digest_size) {
+    if (initialized_) {
+      std::vector<uint8_t> data;
+      data.resize(digest_size);
```
[the remark this answers is not in the document]
Comment (Harrm) [reply src/crypto/hmac_provider/hmac_provider_impl.cpp:185]: Sorry, may I ask why?

File: src/security/noise/crypto/noise_dh.cpp [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    std::copy_n(public_key.begin(), pub.size(), pub.begin());
+    return x25519.dh(priv, pub);
+  }
+
+  int DiffieHellmanImpl::dhSize() const {
+    return 32;
+  }
+
+  std::string DiffieHellmanImpl::dhName() const {
+    return "25519";
```
Comment (Harrm) [inline src/security/noise/crypto/noise_dh.cpp:35]: Why not x25519, is this from noise spec?

Comment (Harrm) [reply src/security/noise/crypto/noise_dh.cpp:35]: I'm just concerned that there's a plenty of things called *25519.

File: test/libp2p/crypto/chachapoly_test.cpp [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+ * @then the decryption result equals to expected
+ */
+TEST_F(ChaChaPolyTest, Decrypt) {
+  ChaCha20Poly1305Impl codec(key);//, Mode::DECRYPT);
+
+  EXPECT_OUTCOME_TRUE(result, codec.decrypt(nonce, ciphertext, aad));
+  ASSERT_EQ(result, plaintext);
+}
+
+//disabled due to broken interface
```
Comment (Harrm) [inline test/libp2p/crypto/chachapoly_test.cpp:159]: Perhaps should be removed.

**Review by Harrm — COMMENTED** [review]
Saw suspiciously few tests in the end as for this gargantuan PR.

File: src/crypto/hmac_provider/hmac_provider_impl.cpp [hunk 17]
```diff
@@ -40,38 +43,156 @@ namespace libp2p::crypto::hmac {
...
+    HMAC_CTX *ctx_copy = HMAC_CTX_new();
+    if (nullptr == ctx_copy) {
+      return HmacProviderError::FAILED_INITIALIZE_CONTEXT;
+    }
+    if (1 != HMAC_CTX_copy(ctx_copy, hmac_ctx_)) {
+      return HmacProviderError::FAILED_INITIALIZE_CONTEXT;
+    }
+    auto free_ctx_copy = gsl::finally([ctx_copy] { HMAC_CTX_free(ctx_copy); });
+    std::vector<uint8_t> result;
+    result.resize(digestSize());
```
[the remark this answers is not in the document]
Comment (iceseer) [reply src/crypto/hmac_provider/hmac_provider_impl.cpp:144]: ```reserve``` unneccessary here. ```resize``` will allocate memory desired size in a single allocation.

File: src/crypto/hmac_provider/hmac_provider_impl.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (igor-egorov) [reply src/crypto/hmac_provider/hmac_provider_impl.cpp:185]: @Harrm Discussed with Kamil offline - no need to invoke reserve here.

File: include/libp2p/crypto/hmac_provider/hmac_provider_impl.hpp [hunk 18]
```diff
@@ -6,13 +6,40 @@
...
 
-#include <libp2p/crypto/hmac_provider.hpp>
+#include <openssl/evp.h>
 #include <gsl/span>
+#include <libp2p/crypto/hmac_provider.hpp>
 
 namespace libp2p::crypto::hmac {
-  class HmacProviderImpl : public HmacProvider {
-    using HashType = common::HashType;
+  class HmacProviderCtrImpl : public HmacProviderCtr {
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply include/libp2p/crypto/hmac_provider/hmac_provider_impl.hpp:14]: There are both interfaces and both classes:
`HmacProviderCtr` and its `Impl`
and
`HmacProvider` and its `Impl`

Non-ctr version is required by already existing code. Ctr was added to support the new code. Old (non-ctr) provider implementation was generalized to use ctr under the hood.

As the resolution - I move the ctr version to its own header then.

File: include/libp2p/crypto/chachapoly.hpp (hunk 4 above)
Comment (igor-egorov) [reply include/libp2p/crypto/chachapoly.hpp:37]: removed. not required anymore

File: include/libp2p/crypto/x25519_provider.hpp (hunk 8 above)
Comment (igor-egorov) [reply include/libp2p/crypto/x25519_provider.hpp:51]: Won't change. 
The reasons are:
(1) this is a common name for a method that does DH keys multiplication. 
(2) the name is preserved to be consistent with go and rust implementations.

File: include/libp2p/security/noise/crypto/message_patterns.hpp (hunk 10 above)
Comment (igor-egorov) [reply include/libp2p/security/noise/crypto/message_patterns.hpp:14]: Nothing to put here. Just an implementation details which are not going to be used outside this particular noise protocol implementation. Transferred from go impl.

File: include/libp2p/security/noise/crypto/state.hpp (hunk 11 above)
Comment (igor-egorov) [reply include/libp2p/security/noise/crypto/state.hpp:34]: Agree. The names was preserved from go implementation. Moving encrypt/decrypt methods out of "states" will require significant redesign. Left it as is for now.

File: include/libp2p/security/noise/handshake.hpp (hunk 12 above)
Comment (igor-egorov) [reply include/libp2p/security/noise/handshake.hpp:28]: Constructor has a numerous of arguments - agree. Since that class is not exposed to dependency injector and is constructed only inside the Noise adapter - this is ok.

File: src/security/noise/crypto/noise_dh.cpp (hunk 15 above)
Comment (igor-egorov) [reply src/security/noise/crypto/noise_dh.cpp:35]: The constant is prescribed by noise spec. https://github.com/libp2p/specs/tree/master/noise#noise-protocol-name
image

Note that XX here stands for handshake pattern name

**Review by Harrm — APPROVED** [review]

**Review by kamilsa — APPROVED** [review]
